modules/sdk-coin-sui/src/lib/transactionBuilder.ts

```diff
@@ -285,31 +280,26 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
    * Validates all fields are defined
    */
   private validateTransactionFields(): void {
-    if (this._type === undefined) {
-      throw new BuildTransactionError('Invalid transaction: missing type');
-    }
-    if (this._sender === undefined) {
-      throw new BuildTransactionError('Invalid transaction: missing sender');
-    }
-    if (this._gasBudget === undefined) {
-      throw new BuildTransactionError('Invalid transaction: missing gas budget');
-    }
-    if (this._gasPrice === undefined || this._gasPrice !== SUI_GAS_PRICE) {
-      throw new BuildTransactionError('Invalid transaction: missing/incorrect gas price');
-    }
-    if (this._payTx === undefined) {
-      throw new BuildTransactionError('Invalid transaction: missing payTx');
-    }
-    if (this._gasPayment === undefined) {
-      throw new BuildTransactionError('Invalid transaction: missing gas payment');
+    assert(this._type, new BuildTransactionError('type is required before building'));
+    assert(this._sender, new BuildTransactionError('sender is required before building'));
+    assert(this._payTx, new BuildTransactionError('payTx is required before building'));
+    assert(this._gasBudget, new BuildTransactionError('gasBudget is required before building'));
+    assert(this._gasPrice, new BuildTransactionError('gasPrice is required before building'));
+    assert(this._gasPayment, new BuildTransactionError('gasPayment is required before building'));
+
+    if (this._gasPrice !== SUI_GAS_PRICE) {
+      throw new BuildTransactionError('Invalid transaction: incorrect gas price');
     }
 
     const coinIds = this._payTx.coins.map((coin) => coin.objectId);
-    if (coinIds.includes(this._gasPayment.objectId)) {
+    if (coinIds.includes(this._gasPayment.objectId) && this._type === SuiTransactionType.Pay) {
       throw new BuildTransactionError(
-        `Invalid gas Payment ${this._gasPayment.objectId}, cannot be one of the inputCoins`
+        `Invalid gas payment ${this._gasPayment.objectId}: cannot be one of the inputCoins`
       );
     }
+    if (coinIds[0] !== this._gasPayment.objectId && this._type !== SuiTransactionType.Pay) {
+      throw new BuildTransactionError(`Invalid gas payment: must use the first input coin as the gas payment`);
+    }
```

Comment:
does it make sense to re-order the input coins so the gasPayment object is always first for PaySui and PayAllSui instead of erroring out?

Reply:
The re-order logic is in bgms, and it should pass the correct inputs to SDK.
But now thinking of this and consider external clients using SDK directly, I should update it to re-order instead of errorring out

Reply:
Updated!